tests/kernel/timer/timer_behavior/src/tick_timer_train.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/time_units.h>
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+
+#define TIMERS 1
+#define TEST_SECONDS 10
+#define MAX_CALLBACKS (CONFIG_SYS_CLOCK_TICKS_PER_SEC*TEST_SECONDS)
+
+struct timer_wrapper {
+	uint32_t callbacks;
+	struct k_timer tm;
+	uint32_t last_isr;
+	int64_t last_scheduled;
+	uint32_t max_delta;
+};
+
+K_SEM_DEFINE(timers_sem, 0, K_SEM_MAX_LIMIT);
+
+struct timer_wrapper timers[TIMERS];
+
+void tm_fn(struct k_timer *tm)
+{
+	struct timer_wrapper *tm_wrap =
+		CONTAINER_OF(tm, struct timer_wrapper, tm);
+	uint32_t now = k_cycle_get_32();
+
+	if (tm_wrap->last_isr != 0) {
+		uint32_t delta = now - tm_wrap->last_isr;
+
+		tm_wrap->max_delta = delta > tm_wrap->max_delta ? delta : tm_wrap->max_delta;
+	}
+	tm_wrap->last_isr = now;
+	tm_wrap->callbacks++;
+	if (tm_wrap->callbacks >= MAX_CALLBACKS) {
+		k_timer_stop(tm);
+		k_sem_give(&timers_sem);
+	} else {
+		int64_t next = tm_wrap->last_scheduled + TIMERS;
+
+		tm_wrap->last_scheduled = next;
+		k_timer_start(tm, K_TIMEOUT_ABS_TICKS(next), K_NO_WAIT);
+	}
+}
+
+
+/**
+ * @brief Test timers can be scheduled 1 tick apart without issues
+ *
+ * Schedules timers with absolute scheduling with a 1 tick
+ * period. Measures the total time elapsed and tries to run
+ * some fake busy work while doing so. If the print outs don't show up or
+ * the timer train is late to the station, the test fails.
+ */
+ZTEST(timer_tick_train, test_one_tick_timer_train)
+{
+	const uint32_t max_time = TEST_SECONDS*1000 + 1000;
+
+	TC_PRINT("Initializing %u Timers, Tick Rate %uHz, Expecting %u callbacks in %u ms\n",
+		 TIMERS, CONFIG_SYS_CLOCK_TICKS_PER_SEC, MAX_CALLBACKS, max_time);
+
+	for (int i = 0; i < TIMERS; i++) {
+		k_timer_init(&timers[i].tm, tm_fn, NULL);
+		timers[i].max_delta = 0;
+	}
+
+	TC_PRINT("Starting Timers with Skews\n");
+	int64_t tick = k_uptime_ticks();
+
+	for (int i = 0; i < TIMERS; i++) {
+		timers[i].last_scheduled = tick+1;
+		k_timer_start(&timers[i].tm, K_TIMEOUT_ABS_TICKS(tick + 1), K_NO_WAIT);
+	}
```

Comment:
Question: Is it worth aligning to a tick boundary prior to starting the timer(s)?

Reply:
Or just skip the analysis for the first pair of times; the timer hooks will self-align by definition anyway.

Reply:
Probably not a big deal, they should self align and the test is try to ensure a mean usefulness with maximal timer scheduling. This won’t change that I don’t think.

I do print out the maximum cycle delta but that’s really not that useful probably, one tempted to remove it.